File: sdks/nodejs/src/kv.ts
import { CloudPoolClient } from './index';

export interface KvEntry {
  key: string;
  value: any;
  createdAt?: string;
  updatedAt?: string;
}

export class KvClient {
  constructor(private readonly client: CloudPoolClient) {}

  public async get<T = unknown>(key: string): Promise<T> {
    const response = await this.client.request(
      'GET',
      `/api/v1/kv/${encodeURIComponent(key)}`
    );
    const json = await response.json();
    return json.value;
  }

  public async set<T = unknown>(
    key: string,
    value: any,
    ttlSeconds?: number
  ): Promise<void> {
    const body: Record<string, any> = { value };
    if (ttlSeconds !== undefined) body.ttlSeconds = ttlSeconds;
    await this.client.request(
      'PUT',
      `/api/v1/kv/${encodeURIComponent(key)}`,
      {
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify(body),
      }
    );
  }

  public async delete<T = unknown>(key: string): Promise<void> {
    await this.client.request(
      'DELETE',
      `/api/v1/kv/${encodeURIComponent(key)}`
    );
  }

  public async list<T = unknown>(prefix?: string): Promise<KvEntry[]> {
    const query = prefix
      ? `?prefix=${encodeURIComponent(prefix)}`
      : '';
    const response = await this.client.request(
      'GET',
      `/api/v1/kv${query}`
    );
    return response.json();
  }
}

File: sdks/go/kv.go
package cloudpool

import "context"

type KVClient struct {
	client *CloudPoolClient
}

type KVEntry struct {
	Key   string `json:"key"`
	Value string `json:"value"`
	TTL   string `json:"ttl,omitempty"`
}

type SetKVRequest struct {
	Value string `json:"value"`
	TTL   int    `json:"ttl,omitempty"`
}

func (k *KVClient) Set(ctx context.Context, key, value string, ttl int) error {
	body := SetKVRequest{Value: value}
	if ttl > 0 {
		body.TTL = ttl
	}
	_, err := k.client.Post(ctx, "/api/kv/"+key, body, nil)
	return err
}

func (k *KVClient) Get(ctx context.Context, key string) (*KVEntry, error) {
	var resp KVEntry
	err := k.client.GetJSON(ctx, "/api/kv/"+key, nil, &resp)
	if err != nil {
		return nil, err
	}
	return &resp, nil
}

func (k *KVClient) List(ctx context.Context) ([]KVEntry, error) {
	var resp []KVEntry
	err := k.client.GetJSON(ctx, "/api/kv", nil, &resp)
	if err != nil {
		return nil, err
	}
	return resp, nil
}

func (k *KVClient) Delete(ctx context.Context, key string) error {
	_, err := k.client.Delete(ctx, "/api/kv/"+key)
	return err
}
